Plan mgmt

gabrielhartmann opened the pull request:
First MVP description of plan management.

mgummelt commented on docs/tutorial.md:
```diff
@@ -349,6 +349,137 @@ If the deployment is complete, `plan.status` in the returned JSON
+are groups of Blocks which have some semantic relationship.  At least
+two plans are generated automatically by the SDK when presented with
+a `ServiceSpecification`.  These are the deployment and recovery plans.  
+The deployment plan is concerned with rolling out service components for
```
what are "service components"?  Can you just say "tasks"?

mgummelt commented on docs/tutorial.md:
```diff
@@ -349,6 +349,137 @@ If the deployment is complete, `plan.status` in the returned JSON
+As noted earlier, stateful services often have particular deployment 
+requirements.  The SDK organizes deployment of tasks by means of Plans. 
+Plans are composed of Phases which are composed of Blocks.  A Block 
+encapsulates the smallest element of work executed in a Plan.  Phases
```
Might be instructive to go ahead and relate a block to a task here.  Since conceptually block == task

gabrielhartmann replied:
See lines 361-362

mgummelt commented on docs/tutorial.md:
```diff
@@ -349,6 +349,137 @@ If the deployment is complete, `plan.status` in the returned JSON
 tasks and three `data` tasks running under the service.
 
+# Plan management
+As noted earlier, stateful services often have particular deployment 
```
Users may better understand this section if you start it with the motivating feature: "you may want to manually step through an install or upgrade for (insert reason here), the SDK creates *plans* on your behalf which allow you to do this"

gabrielhartmann replied:
See lines 357-361

mgummelt commented on docs/tutorial.md:
````diff
@@ -349,6 +349,137 @@ If the deployment is complete, `plan.status` in the returned JSON
+Let us examine the deployment plan of the example data-store service.
+Executing the following command:
+
+```bash
````
To be consistent with the other endpoints in this doc, format this as:

```
GET /v1/plan
```

gabrielhartmann replied:
Resolved.

mgummelt commented on docs/tutorial.md:
```diff
@@ -349,6 +349,137 @@ If the deployment is complete, `plan.status` in the returned JSON
+necessary operations which must be performed when transient or permanent
+failures are encountered.
+
+Let us examine the deployment plan of the example data-store service.
```
After installing, ...

gabrielhartmann replied:
See line 374

mgummelt commented on docs/tutorial.md:
````diff
@@ -349,6 +349,137 @@ If the deployment is complete, `plan.status` in the returned JSON
+```
+
+The first thing to notice is that the `TaskSet`s defined in a
+`ServiceSpecification` are mapped to deployment Phases. The data-store
````
s/Phase/phase

gabrielhartmann replied:
Addressed below.

mgummelt commented on docs/tutorial.md:
```diff
@@ -349,6 +349,137 @@ If the deployment is complete, `plan.status` in the returned JSON
+The first thing to notice is that the `TaskSet`s defined in a
+`ServiceSpecification` are mapped to deployment Phases. The data-store
+example service has `TaskSet`s encapsulating `meta-data` and `data`
+Tasks in that order.  The deployment plan above has the respective
```
s/Task/task et. al.

mgummelt replied:
Unless these are java classes, let's not capitalize them.

gabrielhartmann replied:
Addressed below.

mgummelt commented on docs/tutorial.md:
```diff
@@ -349,6 +349,137 @@ If the deployment is complete, `plan.status` in the returned JSON
+Tasks in that order.  The deployment plan above has the respective
+`meta-data` and `data` Phases.
+
+In its default configuration the 2 `data` Tasks and 3 `meta-data` Tasks
```
comma after "configuration"

mgummelt replied:
s/2/two
s/3/three

(I'll stop grammar nazi-ing now and just let suzanne clean it up later)

mgummelt replied:
Under its default configuration, `data-store` launches...

gabrielhartmann replied:
<img width="498" alt="screen shot 2016-10-10 at 5 32 18 pm" src="https://cloud.githubusercontent.com/assets/567111/19255135/8d7648c8-8f0f-11e6-966c-5cabedb36ec9.png">

mgummelt replied:
I wasn't taking issue with "under" vs "its", but rather that if you write "In its default configuration, ...", then to be gramatically correct, the subject following that clause must be the antecedent of "it", which is in this case `data-store`.

But don't bother with it.  Let's just let suzanne fix it up after.

gabrielhartmann replied:
Resolved.

gabrielhartmann replied:
Resolved

mgummelt commented on docs/tutorial.md:
```diff
@@ -349,6 +349,137 @@ If the deployment is complete, `plan.status` in the returned JSON
+By default all Phases and the Blocks within them are rolled out
+serially.  Each Block must reach a `COMPLETE` state before the
+next Block may be started.  Beyond this automatic behavior, it is often 
+desirable to manually pause/resume a deployment.  The
```
why?

gabrielhartmann replied:
See lines 448-451

mgummelt commented on docs/tutorial.md:
```diff
@@ -349,6 +349,137 @@ If the deployment is complete, `plan.status` in the returned JSON
+`has_decision_point` fields above indicate whether a user has indicated 
+a desire to pause a deployment at a particular Block.
+
+A Plan's execution may be paused or continued by POST commands executed
```
Don't you need some mention of strategies here?  Does the reference framework support strategies?

gabrielhartmann replied:
Not supported yet.

mgummelt replied:
continue/interrupt doesn't seem very useful if the only strategy is INSTALL.  If I want to manage my deployment, I probably want to deploy with a non-automatic strategy.

gabrielhartmann replied:
You're right.  Documenting what exists right now.

mgummelt replied:
Let's wait to merge this until the feature exists.  Is there a JIRA?

gabrielhartmann replied:
No.  It does exist on Install plans.  It's just that deployments aren't stopped by default.  They start rolling out immediately.  You can stop it if you like.

gabrielhartmann commented:
@mgummelt:  Addressed comments.  I think we _should_ capitalize Task, Phase, Block etc.  Those words have meanings in the real world.  Capitalization indicates that they are names with non-standard meanings here.  Apart from that, pretty much all of those are Java classes/interfaces anyway.

mgummelt commented:
I disagree, but I'll leave it up to @sascala

mgummelt commented on docs/tutorial.md:
````diff
@@ -349,6 +349,155 @@ If the deployment is complete, `plan.status` in the returned JSON
+data task `data-0` one would issue the following command:
+
+```bash
+curl -X POST -H "Authorization: token=$(dcos config show core.dcos_acs_token)" \
````
This format for API requests is still inconsistent with the rest of the doc

gabrielhartmann commented:
@mgummelt:  Please review and see if you have additional comments.

mgummelt commented:
cc @sascala for editing
